### PR TITLE
[SPARK-35390][SQL] Handle type coercion when resolving V2 functions

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

Handle type coercion when resolving V2 function. In particular:
- prior to evaluating function arguments, insert cast whenever the argument type doesn't match the expected input type.
- use `BoundFunction.inputTypes()` to lookup magic method for scalar function

### Why are the changes needed?

For V2 functions, the actual argument types should not necessarily match those of the input types, and Spark should handle type coercion whenever it is needed.

### Does this PR introduce _any_ user-facing change?

Yes. Now V2 function resolution should be able to handle type coercion properly.

### How was this patch tested?

Added a few new tests.

### CODE REVIEW

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java [hunk 1]
```diff
@@ -117,6 +120,8 @@
  *   <li>{@link org.apache.spark.sql.types.DateType}: {@code int}</li>
  *   <li>{@link org.apache.spark.sql.types.TimestampType}: {@code long}</li>
  *   <li>{@link org.apache.spark.sql.types.BinaryType}: {@code byte[]}</li>
+ *   <li>{@link org.apache.spark.sql.types.CalendarIntervalType}:
```
Comment (sunchao) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java:123]: this is not related - lemme know if I should use a separate PR.

**Comment by sunchao** [comment]
cc @cloud-fan @rdblue @viirya @dongjoon-hyun

**Comment by viirya** [comment]
Thanks @sunchao!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -2169,12 +2169,29 @@ class Analyzer(override val catalogManager: CatalogManager)
...
+                    if (arg.dataType != ty) {
+                      if (Cast.canCast(arg.dataType, ty)) {
+                        Cast(arg, ty)
+                      } else {
+                        throw QueryCompilationErrors.v2FunctionCastError(bound, arg, ty)
+                      }
+                    } else {
+                      arg
+                    }
+                  }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2187]: As `bound` is come from binding on `UnboundFunction`. From the doc of `bind` method, it works like checking input types.

So that's said, when we bind it, it already checks argument data types, why we need another check and cast here?

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @sunchao .

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java:123]: I'm fine. :)

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala [hunk 3]
```diff
@@ -21,15 +21,15 @@ import java.util
...
+import test.org.apache.spark.sql.connector.catalog.functions.JavaLongAdd.{JavaLongAddDefault, JavaLongAddMagic, JavaLongAddMismatchMagic, JavaLongAddStaticMagic}
 import test.org.apache.spark.sql.connector.catalog.functions.JavaStrLen._
 
 import org.apache.spark.SparkException
 import org.apache.spark.sql.{AnalysisException, Row}
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.CodegenObjectFactoryMode.{FALLBACK, NO_CODEGEN}
 import org.apache.spark.sql.connector.catalog.{BasicInMemoryTableCatalog, Identifier, InMemoryCatalog, SupportsNamespaces}
-import org.apache.spark.sql.connector.catalog.functions._
+import org.apache.spark.sql.connector.catalog.functions.{AggregateFunction, _}
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala:32]: ?

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala [hunk 4]
```diff
@@ -312,6 +344,27 @@ class DataSourceV2FunctionSuite extends DatasourceV2SQLBase {
...
   }
 
+  test("SPARK-35390: aggregate function w/ type coercion") {
+    import testImplicits._
+
+    withTable("t1", "t2") {
+      addFunction(Identifier.of(Array("ns"), "avg"), UnboundDecimalAverage)
+
+      (1 to 100).toDF().write.saveAsTable("testcat.ns.t1")
+      checkAnswer(sql(s"SELECT testcat.ns.avg(value) from testcat.ns.t1"),
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala:354]: nit. `s"` -> `"`.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala:354]: And, ditto for line 358 and 362.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala [hunk 5]
```diff
@@ -466,6 +526,60 @@ class DataSourceV2FunctionSuite extends DatasourceV2SQLBase {
...
+      // put interval type here for testing purpose
+      inputType.fields(0).dataType match {
+        case _: NumericType | _: DayTimeIntervalType => DecimalAverage
+        case dataType =>
+          throw new UnsupportedOperationException(s"Unsupported input type: $dataType")
+      }
+    }
+
+    override def description(): String =
+      """iavg: produces an average using decimal division, ignoring nulls
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala:546]: Just a question. What is the meaning of `i` here? Previously, I thought it stands for `Integer`.
BTW, `davg` may look a little confusing because it can be read as DecimalAverage or DoubleAverage.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala [hunk 6]
```diff
@@ -466,6 +526,60 @@ class DataSourceV2FunctionSuite extends DatasourceV2SQLBase {
...
+        |  iavg(integral) -> decimal
+        |  iavg(float) -> decimal
+        |  iavg(decimal) -> decimal""".stripMargin
+  }
+
+  object DecimalAverage extends AggregateFunction[(Decimal, Int), Decimal] {
+    val PRECISION: Int = 15
+    val SCALE: Int = 5
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala:554]: Sorry, but may I ask where these (precision 15, scale 5) came from? I might miss the context, but this looks like not non-conventional values in Apache Spark codebase. Is it intentionally chosen for some purpose?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2187]: Yes `bind` checks input types, and potentially allows multiple combinations of them. However, when evaluating the UDF, especially when dealing with magic method, Spark only accept a single set of input parameter types, and so we'll need to insert cast if necessary.

For instance, a UDF can accept both `int` and `decimal` as input types in `bind`, and implements magic method using `decimal` parameters. Spark then should cast `int` arguments to `decimal` when necessary.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala (hunk 5 above)
Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala:546]: my bad - let me fix the name to be something more meaningful, like `decimal_avg`.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala (hunk 6 above)
Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala:554]: sorry this is a little confusing - I was using this to test some negative case like passing `long` which is `decimal(20, 0)` to this `decimal(15, 5)` which won't work without losing precision. But, I no longer need this now. Will remove.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2187]: Since we have two places of type coercion, is there any possibility of bug where we do differently in some cases at both places?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2187]: Thanks for explaining it. So that is said, `bind` can return a implementation with magic method, which takes decimal input, when Spark binds it with IntegerType input?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2187]: Then the question is, how the UDF knows Spark can handle the type coercion case?

Using above example, when Spark binds it with IntegerType, the UDF must know Spark can cast int to decimal, so it can return an implementation with magic method taking decimal input?

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2187]: > Since we have two places of type coercion, is there any possibility of bug where we do differently in some cases at both places?

Hmm @dongjoon-hyun what are the two places of type coercion? do you mean the potential inconsistency from `bind` and `inputTypes`?

> Thanks for explaining it. So that is said, bind can return a implementation with magic method, which takes decimal input, when Spark binds it with IntegerType input?

Yes correct. Spark will insert cast `int -> decimal`. 

Also from the Java doc of `inputTypes`: "If the types returned differ from the types passed to `bind(StructType)`, Spark will cast input values to the required data types. This allows implementations to delegate input value casting to Spark."

> Using above example, when Spark binds it with IntegerType, the UDF must know Spark can cast int to decimal, so it can return an implementation with magic method taking decimal input?

Yes the implementor of the UDF should know whether such a cast is valid. Spark also checks this during analysis and throws error if not.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala (hunk 3 above)
Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2FunctionSuite.scala:32]: You mean to split into two lines?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java:123]: I'm wondering if we should forbid the old interval type in v2 function as it's a new feature. cc @MaxGekk

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2187]: It seems to me that a more natural approach is: the v2 function tells Spark all the overloads of its function (e.g. `substring(string)` and `substring(binary)`) , then Spark decides which overload to call and apply necessary type coercions.

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2187]: @cloud-fan I think this approach will not allow us to handle function with varargs, for instance: `max(int, long, float, double, int, long, long, double)`. For this scenario, `bind` can check the input arguments and return a proper implementation of the function (although we still need to do something extra for varargs in Spark).

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java (hunk 1 above)
Comment (sunchao) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java:123]: ohh I'm happy to remove it then if there is any concern. We can add it separately later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2187]: makes sense

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -2169,12 +2169,29 @@ class Analyzer(override val catalogManager: CatalogManager)
...
                   }
 
+                  if (bound.inputTypes().length != arguments.length) {
+                    throw QueryCompilationErrors.v2FunctionInvalidInputTypeLengthError(
+                      bound, arguments)
+                  }
+
+                  val castedArguments = arguments.zip(bound.inputTypes()).map { case (arg, ty) =>
+                    if (arg.dataType != ty) {
+                      if (Cast.canCast(arg.dataType, ty)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2179]: This is a bit hacky to do type coercion manually, instead of using the existing type coercion framework. e.g. how do you support ANSI type coercion here? See `AnsiTypeCoercion` for more details.

My proposal is to let `Invoke`/`StaticInvoke` extend `ImplicitCastInputTypes`, and leverage the existing type coercion framework.

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2179]: Good point! I totally ignored `AnsiCast`. Let me try out your suggestion. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -2189,7 +2194,7 @@ class Analyzer(override val catalogManager: CatalogManager)
 
     private def processV2ScalarFunction(
         scalarFunc: ScalarFunction[_],
-        inputType: StructType,
+        argumentTypes: StructType,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2197]: do we still need this parameter? we can simplify the scala function branch
```
scalarFunc.getClass.getDeclaredMethod("produceResult", classOf[InternalRow]) ...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 9]
```diff
@@ -231,6 +236,11 @@ object SerializerSupport {
  * @param dataType The expected return type of the function call
  * @param functionName The name of the method to call.
  * @param arguments An optional list of expressions to pass as arguments to the function.
+ * @param methodInputTypes An optional list of data types specifying the input types for the method
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:239]: Shall we use `inputTypes: Seq[AbstractDataType]` directly? Nil means no implicit cast needed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 10]
```diff
@@ -1470,6 +1470,18 @@ private[spark] object QueryCompilationErrors {
...
   }
 
+  def v2FunctionInvalidInputTypeLengthError(
+      bound: BoundFunction,
+      args: Seq[Expression]): Throwable = {
+    new AnalysisException(s"Invalid bound function '${bound.name()}: there are ${args.length} " +
+        s"arguments but ${bound.inputTypes().length} parameters returned from 'inputTypes()'")
+  }
+
+  def v2FunctionCastError(bound: BoundFunction, arg: Expression, ty: DataType): Throwable = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1480]: this is no longer needed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 9 above)
Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:239]: Oops you are right - didn't notice the `Nil` case. We can't do this for `Invoke` though since its children include the `targetObject`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 8 above)
Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yep no longer needed. Will remove.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by sunchao** [comment]
Thanks @cloud-fan @dongjoon-hyun and @viirya !
